platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileCompactor.java

```diff
@@ -0,0 +1,470 @@
+        snapshotCompactionLock.acquireUninterruptibly();
+        // Check if compaction is currently in progress. If so, flush and close the current file, so
+        // it's included to the snapshot
+        final DataFileWriter<?> compactionWriter = currentCompactionWriter.get();
+        if (compactionWriter != null) {
+            compactionWasInProgress.set(true);
+            compactionLevelInProgress.set(compactionWriter.getMetadata().getCompactionLevel());
+            finishCurrentCompactionFile();
+            // Don't start a new compaction file here, as it would be included to snapshots, but
+            // it shouldn't, as it isn't fully written yet. Instead, a new file will be started
+            // right after snapshot is taken, in resumeCompaction()
+        }
+        // Don't release the lock here, it will be done later in resumeCompaction(). If there is no
+        // compaction currently running, the lock will prevent starting a new one until snapshot is
+        // done
+    }
+
+    /**
+     * Resumes compaction previously put on hold with {@link #pauseCompaction()}. If there was no
+     * compaction running at that moment, but new compaction was started (and blocked) since {@link
+     * #pauseCompaction()}, this new compaction is resumed.
+     *
+     * <b>This method must be always balanced with and called after {@link #pauseCompaction()}. If
+     * there are more / less calls to resume compactions than to pause, or if they are called in a
+     * wrong order, it will result in deadlocks.</b>
+     *
+     * @throws IOException If an I/O error occurs
+     */
+    public void resumeCompaction() throws IOException {
+        try {
+            if (compactionWasInProgress.getAndSet(false)) {
+                assert currentCompactionWriter.get() == null;
+                assert currentCompactionReader.get() == null;
+                startNewCompactionFile(compactionLevelInProgress.getAndSet(0));
+            }
+        } finally {
+            snapshotCompactionLock.release();
+        }
+    }
+
+    /**
+     * Compact data files in the collection according to the compaction algorithm.
+     *
+     * @param reportDurationMetricFunction   function to report how long compaction took, in ms
+     * @param reportSavedSpaceMetricFunction function to report how much space was compacted, in Mb
+     * @throws IOException          if there was a problem merging
+     * @throws InterruptedException if the merge thread was interrupted
+     * @return true if compaction was performed, false otherwise
+     */
+    @SuppressWarnings("unchecked")
+    public boolean compact(
+            CASableLongIndex index,
+            @Nullable final BiConsumer<Integer, Long> reportDurationMetricFunction,
+            @Nullable final BiConsumer<Integer, Double> reportSavedSpaceMetricFunction)
+            throws IOException, InterruptedException {
+
+        final String storeName = dataFileCollection.getStoreName();
```

Comment:
Store name can be a field in this compactor, so you don't have to expose `getStoreName()` in `DataFileCollection`.